File: Grunt/Grunt/Models/HaloInfinite/Csr.cs
// <copyright file="Csr.cs" company="Den Delimarsky">
// Developed by Den Delimarsky.
// Den Delimarsky licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.
// The underlying API powering Grunt is managed by 343 Industries and Microsoft. This wrapper is not endorsed by 343 Industries or Microsoft.
// </copyright>

namespace OpenSpartan.Grunt.Models.HaloInfinite
{
    /// <summary>
    /// A record of a match Competitive Skill Rank (CSR).
    /// </summary>
    [IsAutomaticallySerializable]
    public class Csr
    {
        /// <summary>
        /// Gets or sets the CSR value.
        /// </summary>
        public int Value { get; set; }

        /// <summary>
        /// Gets or sets the number of matches remaining to measure the CSR.
        /// </summary>
        public int MeasurementMatchesRemaining { get; set; }

        /// <summary>
        /// Gets or sets the tier.
        /// </summary>
        public string? Tier { get; set; }

        /// <summary>
        /// Gets or sets the tier start.
        /// </summary>
        public int TierStart { get; set; }

        /// <summary>
        /// Gets or sets the subtier.
        /// </summary>
        public int SubTier { get; set; }

        /// <summary>
        /// Gets or sets the next tier.
        /// </summary>
        public string? NextTier { get; set; }

        /// <summary>
        /// Gets or sets the start of the next tier.
        /// </summary>
        public int NextTierStart { get; set; }

        /// <summary>
        /// Gets or sets the next subtier.
        /// </summary>
        public int NextSubTier { get; set; }

        /// <summary>
        /// Gets or sets the number of initial measurement matches.
        /// </summary>
        public int InitialMeasurementMatches { get; set; }
    }
}

File: src/node/src/models/halo-infinite/skill/csr.ts
/**
 * Competitive Skill Rank (CSR) information.
 *
 * CSR is the ranked rating system in Halo Infinite multiplayer.
 */
export interface Csr {
  /** Current CSR value */
  Value?: number;
  /** Measurement matches remaining before placement */
  MeasurementMatchesRemaining?: number;
  /** Current tier name (e.g., "Diamond", "Onyx") */
  Tier?: string;
  /** CSR value at start of current tier */
  TierStart?: number;
  /** Current sub-tier within the tier (1-6) */
  SubTier?: number;
  /** Next tier name */
  NextTier?: string;
  /** CSR value at start of next tier */
  NextTierStart?: number;
  /** Next sub-tier */
  NextSubTier?: number;
  /** Initial number of placement matches required */
  InitialMeasurementMatches?: number;
  /** Initial demotion protection matches */
  InitialDemotionProtectionMatches?: number;
  /** Remaining demotion protection matches */
  DemotionProtectionMatchesRemaining?: number;
}

/**
 * Skill result from a match.
 */
export interface SkillResult {
  /** Result data */
  Result?: unknown;
}

/**
 * Container for match skill results.
 */
export interface MatchSkillInfo {
  /** Skill results for each player */
  Value?: SkillResult[];
}

/**
 * Playlist CSR container for a single playlist.
 */
export interface PlaylistCsrContainer {
  /** Current CSR */
  Current?: Csr;
  /** Season max CSR */
  SeasonMax?: Csr;
  /** All-time max CSR */
  AllTimeMax?: Csr;
}

/**
 * Container for playlist CSR results.
 */
export interface PlaylistCsrResultContainer {
  /** CSR results */
  Value?: unknown[];
}
